tests/src/test/scala/akka/kafka/scaladsl/RebalanceExtSpec.scala

```diff
@@ -0,0 +1,427 @@
+    (topicMap, topicPartitionFutureMap, topicIdxMap, topicMap.keySet, msgTpMap, producers, messageStoreAndAck)
+  }
+
+  // publish a numeric series tagged messages across all topic-partitions
+  // e.g., for 18 messages:  messageId(1 to 9) => topic-1-1-0; messageId(10 to 18) => topic-1-1-1
+  def publishMessages(
+      topicCount: Int,
+      partitionCount: Int,
+      perPartitionMessageCount: Int,
+      topicIdxMap: Map[Int, String]
+  ): (Map[Int, String], Seq[Future[Done]], Map[Int, (AtomicInteger, Promise[Done], Promise[Done])]) = {
+    var msgTpMap = Map[Int, String]()
+    var messageStoreAndAck = Map[Int, (AtomicInteger, Promise[Done], Promise[Done])]()
+    val producers: Seq[Future[Done]] = topicIdxMap.keySet
+      .map { topicIdx =>
+        val topic1 = topicIdxMap(topicIdx)
+        val topicOffset = (topicIdx - 1) * partitionCount * perPartitionMessageCount
+        (0 until partitionCount).toSet.map { partitionIdx: Int =>
+          val startMessageIdx = partitionIdx * perPartitionMessageCount + 1 + topicOffset
+          val endMessageIdx = startMessageIdx + perPartitionMessageCount - 1
+          val messageRange = startMessageIdx to endMessageIdx
+          messageRange.foreach(messageId => {
+            val topicPartition = s"$topic1-$partitionIdx"
+            msgTpMap = msgTpMap.updated(messageId, topicPartition)
+            messageStoreAndAck =
+              messageStoreAndAck.updated(messageId, (new AtomicInteger(0), Promise[Done], Promise[Done]))
+          })
+          produce(topic1, messageRange, partitionIdx).map { f =>
+            val topicPartition = s"$topic1-$partitionIdx"
+            log.debug(s"Produced messages $startMessageIdx to $endMessageIdx to topicPartition $topicPartition")
+            f
+          }
+        }.toSeq
+      }
+      .toSeq
+      .flatten
+    (msgTpMap, producers, messageStoreAndAck)
+  }
+
+  def businessFlow1(clientId: String,
+                    perPartitionMessageCount: Int,
+                    topicPartitionFutureMap: Map[String, Promise[Done]],
+                    messageStoreAndAck: Map[Int, (AtomicInteger, Promise[Done], Promise[Done])],
+                    sharedKillSwitch: SharedKillSwitch,
+                    logPrefix: String): Flow[CommittableMessage[String, String], CommittableOffset, NotUsed] = {
+    Flow.fromFunction { message =>
+      val messageVal = message.record.value.toInt
+      val duplicateCount = messageStoreAndAck(messageVal)._1.incrementAndGet()
+      val topicPartition = s"${message.record.topic}-${message.record.partition}"
+      val msg1 =
+        s"$logPrefix::businessFlow1:offset=${message.committableOffset.partitionOffset.offset} messageId=$messageVal topicPartition=${message.record.topic}-${message.record.partition} consumerId=$clientId duplicateCount=$duplicateCount"
+      if (duplicateCount > 1) {
+        log.warn(s"businessFlow1:received:$msg1")
```

Comment:
Should this say something about a duplicate?

Reply:
done